lib/Basic/Diagnostic.cpp

```
@@ -0,0 +1,119 @@
+    if (!msg.getLocation().isValid()) {
+        return;
+    }
+
+    // Format the location information
+    os << _sourceManager.getBufferName(msg.getLocation()) << ":"
+       << _sourceManager.getSpellingLineNumber(msg.getLocation()) << ":"
+       << _sourceManager.getSpellingColumnNumber(msg.getLocation()) << ": ";
+
+    // Format the severity prefix with colors if supported
+    switch (msg.getSeverity()) {
+    case DiagnosticSeverity::Note: os << "note: "; break;
+    case DiagnosticSeverity::Warning: os << "warning: "; break;
+    case DiagnosticSeverity::Error: os << "error: "; break;
+    case DiagnosticSeverity::Fatal: os << "fatal error: "; break;
+    }
+
+    // Print the actual message
+    os << msg.getMessage() << "\n";
+
+    if (!_sourceManager.getBufferData(msg.getLocation()).empty()) {
+        size_t lineStart = _sourceManager.getLineStart(msg.getLocation());
+        size_t lineEnd = _sourceManager.getLineEnd(msg.getLocation());
+        size_t column
+            = _sourceManager.getSpellingColumnNumber(msg.getLocation());
+
+        // Print the line of source code
+        os << _sourceManager.getBufferData(msg.getLocation())
+                  .slice(lineStart, lineEnd)
+                  .str()
+           << "\n";
+
+        // Print a caret (^) pointing to the specific column
+        os.indent(column - 1);
+        os << "^\n";
+    }
+}
+
+void Diagnostic::printAll(llvm::raw_ostream &os) const
+{
+    for (auto const &msg : _messages) {
+        printDiagnostic(os, msg);
+    }
+
+    // Print summary counts
+    size_t errorCount = 0;
+    size_t warningCount = 0;
+    size_t noteCount = 0;
+
+    for (auto const &msg : _messages) {
+        switch (msg.getSeverity()) {
+        case DiagnosticSeverity::Note: ++noteCount; break;
+        case DiagnosticSeverity::Warning: ++warningCount; break;
+        case DiagnosticSeverity::Error:
+        case DiagnosticSeverity::Fatal: ++errorCount; break;
+        default: break;
```

Useless default would hide the warning if we add a new severity (-Wswitch)